File: examples/lwjgl-map/src/main/java/org/maplibre/nativeffi/examples/lwjglmap/InputController.java
package org.maplibre.nativeffi.examples.lwjglmap;

import static org.lwjgl.glfw.GLFW.GLFW_KEY_0;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_A;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_D;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_DOWN;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_E;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_EQUAL;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_LEFT;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_LEFT_BRACKET;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_MINUS;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_PAGE_DOWN;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_PAGE_UP;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_Q;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_RIGHT;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_RIGHT_BRACKET;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_S;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_UP;
import static org.lwjgl.glfw.GLFW.GLFW_KEY_W;
import static org.lwjgl.glfw.GLFW.GLFW_MOD_CONTROL;
import static org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_LEFT;
import static org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_RIGHT;
import static org.lwjgl.glfw.GLFW.GLFW_PRESS;
import static org.lwjgl.glfw.GLFW.GLFW_RELEASE;
import static org.lwjgl.glfw.GLFW.GLFW_REPEAT;
import static org.lwjgl.glfw.GLFW.glfwSetCursorPosCallback;
import static org.lwjgl.glfw.GLFW.glfwSetKeyCallback;
import static org.lwjgl.glfw.GLFW.glfwSetMouseButtonCallback;
import static org.lwjgl.glfw.GLFW.glfwSetScrollCallback;

import org.maplibre.nativeffi.camera.AnimationOptions;
import org.maplibre.nativeffi.camera.CameraOptions;
import org.maplibre.nativeffi.geo.ScreenPoint;
import org.maplibre.nativeffi.map.MapHandle;

final class InputController implements AutoCloseable {
  private static final double DRAG_ROTATE_FACTOR = 0.5;
  private static final double DRAG_PITCH_FACTOR = 0.5;
  private static final double KEYBOARD_PAN = 120.0;
  private static final double KEYBOARD_ZOOM = 1.25;
  private static final double KEYBOARD_BEARING = 10.0;
  private static final double KEYBOARD_PITCH = 5.0;
  private static final AnimationOptions KEYBOARD_ANIMATION = new AnimationOptions().durationMs(160);
  private static final AnimationOptions RESET_ANIMATION = new AnimationOptions().durationMs(220);

  private final long window;
  private final MapHandle map;
  private boolean leftDown;
  private boolean rightDown;
  private boolean ctrlDown;
  private double lastX;
  private double lastY;
  private double cursorX;
  private double cursorY;

  InputController(long window, MapHandle map) {
    this.window = window;
    this.map = map;
    installCallbacks();
  }

  static void printControls() {
    System.out.println("Controls:");
    System.out.println("  left drag: pan");
    System.out.println("  right drag or Ctrl+left drag: rotate with X, pitch with Y");
    System.out.println("  scroll: zoom at cursor");
    System.out.println("  arrows or WASD: pan");
    System.out.println("  + / -: zoom at center");
    System.out.println("  Q / E: rotate");
    System.out.println("  PageUp / PageDown or [ / ]: pitch");
    System.out.println("  0: reset pitch and bearing");
  }

  private void installCallbacks() {
    glfwSetCursorPosCallback(window, (ignored, x, y) -> onCursor(x, y));
    glfwSetMouseButtonCallback(
        window, (ignored, button, action, mods) -> onMouse(button, action, mods));
    glfwSetScrollCallback(window, (ignored, xOffset, yOffset) -> onScroll(yOffset));
    glfwSetKeyCallback(window, (ignored, key, scancode, action, mods) -> onKey(key, action, mods));
  }

  private void onCursor(double x, double y) {
    cursorX = x;
    cursorY = y;
    var dx = x - lastX;
    var dy = y - lastY;
    lastX = x;
    lastY = y;
    if (rightDown || (leftDown && ctrlDown)) {
      setBearing(currentBearing() + dx * DRAG_ROTATE_FACTOR, false);
      setPitch(currentPitch() - dy * DRAG_PITCH_FACTOR, false);
    } else if (leftDown) {
      map.moveBy(dx, dy);
    }
  }

  private void onMouse(int button, int action, int mods) {
    ctrlDown = (mods & GLFW_MOD_CONTROL) != 0;
    if (button == GLFW_MOUSE_BUTTON_LEFT) {
      leftDown = action == GLFW_PRESS ? true : action == GLFW_RELEASE ? false : leftDown;
    } else if (button == GLFW_MOUSE_BUTTON_RIGHT) {
      rightDown = action == GLFW_PRESS ? true : action == GLFW_RELEASE ? false : rightDown;
    }
    if (action == GLFW_PRESS) {
      map.cancelTransitions();
    }
  }

  private void onScroll(double yOffset) {
    // GLFW reports OS-adjusted scroll deltas; use them directly so trackpads with natural
    // scrolling behave like the host platform expects.
    var scale = Math.pow(2.0, yOffset * 0.25);
    map.scaleBy(scale, new ScreenPoint(cursorX, cursorY));
  }

  private void onKey(int key, int action, int mods) {
    ctrlDown = (mods & GLFW_MOD_CONTROL) != 0;
    if (action != GLFW_PRESS && action != GLFW_REPEAT) {
      return;
    }
    switch (key) {
      case GLFW_KEY_LEFT, GLFW_KEY_A -> map.moveByAnimated(KEYBOARD_PAN, 0.0, KEYBOARD_ANIMATION);
      case GLFW_KEY_RIGHT, GLFW_KEY_D -> map.moveByAnimated(-KEYBOARD_PAN, 0.0, KEYBOARD_ANIMATION);
      case GLFW_KEY_UP, GLFW_KEY_W -> map.moveByAnimated(0.0, KEYBOARD_PAN, KEYBOARD_ANIMATION);
      case GLFW_KEY_DOWN, GLFW_KEY_S -> map.moveByAnimated(0.0, -KEYBOARD_PAN, KEYBOARD_ANIMATION);
      case GLFW_KEY_EQUAL -> map.scaleByAnimated(KEYBOARD_ZOOM, KEYBOARD_ANIMATION);
      case GLFW_KEY_MINUS -> map.scaleByAnimated(1.0 / KEYBOARD_ZOOM, KEYBOARD_ANIMATION);
      case GLFW_KEY_Q -> setBearing(currentBearing() - KEYBOARD_BEARING, true);
      case GLFW_KEY_E -> setBearing(currentBearing() + KEYBOARD_BEARING, true);
      case GLFW_KEY_PAGE_UP, GLFW_KEY_RIGHT_BRACKET ->
          setPitch(currentPitch() + KEYBOARD_PITCH, true);
      case GLFW_KEY_PAGE_DOWN, GLFW_KEY_LEFT_BRACKET ->
          setPitch(currentPitch() - KEYBOARD_PITCH, true);
      case GLFW_KEY_0 -> map.easeTo(new CameraOptions().bearing(0.0).pitch(0.0), RESET_ANIMATION);
      default -> {}
    }
  }

  private double currentBearing() {
    var camera = map.camera();
    return camera.hasBearing() ? camera.bearing() : 0.0;
  }

  private double currentPitch() {
    var camera = map.camera();
    return camera.hasPitch() ? camera.pitch() : 0.0;
  }

  private void setBearing(double bearing, boolean animated) {
    var camera = new CameraOptions().bearing(bearing);
    if (animated) {
      map.easeTo(camera, KEYBOARD_ANIMATION);
    } else {
      map.jumpTo(camera);
    }
  }

  private void setPitch(double pitch, boolean animated) {
    var clamped = Math.max(0.0, Math.min(60.0, pitch));
    var camera = new CameraOptions().pitch(clamped);
    if (animated) {
      map.easeTo(camera, KEYBOARD_ANIMATION);
    } else {
      map.jumpTo(camera);
    }
  }

  @Override
  public void close() {
    glfwSetCursorPosCallback(window, null);
    glfwSetMouseButtonCallback(window, null);
    glfwSetScrollCallback(window, null);
    glfwSetKeyCallback(window, null);
  }
}

File: examples/lwjgl-map/src/main/kotlin/org/maplibre/nativeffi/examples/lwjglmap/InputController.kt
package org.maplibre.nativeffi.examples.lwjglmap

import kotlin.math.pow
import org.lwjgl.glfw.GLFW.GLFW_KEY_0
import org.lwjgl.glfw.GLFW.GLFW_KEY_A
import org.lwjgl.glfw.GLFW.GLFW_KEY_D
import org.lwjgl.glfw.GLFW.GLFW_KEY_DOWN
import org.lwjgl.glfw.GLFW.GLFW_KEY_E
import org.lwjgl.glfw.GLFW.GLFW_KEY_EQUAL
import org.lwjgl.glfw.GLFW.GLFW_KEY_LEFT
import org.lwjgl.glfw.GLFW.GLFW_KEY_LEFT_BRACKET
import org.lwjgl.glfw.GLFW.GLFW_KEY_MINUS
import org.lwjgl.glfw.GLFW.GLFW_KEY_Q
import org.lwjgl.glfw.GLFW.GLFW_KEY_RIGHT
import org.lwjgl.glfw.GLFW.GLFW_KEY_RIGHT_BRACKET
import org.lwjgl.glfw.GLFW.GLFW_KEY_S
import org.lwjgl.glfw.GLFW.GLFW_KEY_UP
import org.lwjgl.glfw.GLFW.GLFW_KEY_W
import org.lwjgl.glfw.GLFW.GLFW_MOD_CONTROL
import org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_LEFT
import org.lwjgl.glfw.GLFW.GLFW_MOUSE_BUTTON_RIGHT
import org.lwjgl.glfw.GLFW.GLFW_PRESS
import org.lwjgl.glfw.GLFW.GLFW_RELEASE
import org.lwjgl.glfw.GLFW.GLFW_REPEAT
import org.lwjgl.glfw.GLFW.glfwSetCursorPosCallback
import org.lwjgl.glfw.GLFW.glfwSetKeyCallback
import org.lwjgl.glfw.GLFW.glfwSetMouseButtonCallback
import org.lwjgl.glfw.GLFW.glfwSetScrollCallback
import org.maplibre.nativeffi.geo.ScreenPoint

/**
 * Decodes GLFW input into camera commands.
 *
 * GLFW delivers these callbacks on the render loop thread, which does not own the map, so this only
 * produces commands; the runtime loop applies them on the map's thread. Anything needing the
 * current viewport is converted here, where the viewport lives.
 */
internal class InputController(
  private val window: Long,
  private val commands: CommandQueue,
  private val renderRequest: RenderRequest,
  private val viewport: () -> Viewport,
) : AutoCloseable {
  private var leftDown = false
  private var rightDown = false
  private var ctrlDown = false
  private var lastX = 0.0
  private var lastY = 0.0
  private var cursorX = 0.0
  private var cursorY = 0.0

  init {
    glfwSetCursorPosCallback(window) { _, x, y -> onCursor(x, y) }
    glfwSetMouseButtonCallback(window) { _, button, action, mods -> onMouse(button, action, mods) }
    glfwSetScrollCallback(window) { _, _, yOffset -> onScroll(yOffset) }
    glfwSetKeyCallback(window) { _, key, _, action, mods -> onKey(key, action, mods) }
  }

  private fun onCursor(x: Double, y: Double) {
    cursorX = x
    cursorY = y
    val dx = x - lastX
    val dy = y - lastY
    lastX = x
    lastY = y
    if (dx == 0.0 && dy == 0.0) {
      return
    }
    if (rightDown || (leftDown && ctrlDown)) {
      commands.push(CameraCommand.AdjustBearing(dx * DRAG_ROTATE_FACTOR))
      commands.push(CameraCommand.PitchBy(dy * DRAG_PITCH_FACTOR))
    } else if (leftDown) {
      commands.push(CameraCommand.MoveBy(dx, dy))
    } else {
      return
    }
    renderRequest.set()
  }

  private fun onMouse(button: Int, action: Int, mods: Int) {
    ctrlDown = (mods and GLFW_MOD_CONTROL) != 0
    val wasDragging = dragging
    when (button) {
      GLFW_MOUSE_BUTTON_LEFT ->
        leftDown =
          if (action == GLFW_PRESS) true else if (action == GLFW_RELEASE) false else leftDown

      GLFW_MOUSE_BUTTON_RIGHT ->
        rightDown =
          if (action == GLFW_PRESS) true else if (action == GLFW_RELEASE) false else rightDown
    }
    if (action == GLFW_PRESS) {
      // Queued ahead of the drag's own commands, so the transition stops before the first delta
      // lands.
      commands.push(CameraCommand.CancelTransitions)
    }
    // The deltas in between belong to one live gesture, so the map hears about the gesture rather
    // than a stream of unrelated camera commands.
    if (dragging != wasDragging) {
      commands.push(CameraCommand.SetGestureInProgress(dragging))
    }
  }

  private val dragging: Boolean
    get() = leftDown || rightDown

  private fun onScroll(yOffset: Double) {
    // GLFW reports OS-adjusted scroll deltas; use them directly so trackpads with natural
    // scrolling behave like the host platform expects.
    val scale = 2.0.pow(yOffset * 0.25)
    commands.push(CameraCommand.ScaleBy(scale, ScreenPoint(cursorX, cursorY)))
    renderRequest.set()
  }

  private fun onKey(key: Int, action: Int, mods: Int) {
    ctrlDown = (mods and GLFW_MOD_CONTROL) != 0
    if (action != GLFW_PRESS && action != GLFW_REPEAT) {
      return
    }
    val command =
      when (key) {
        GLFW_KEY_LEFT,
        GLFW_KEY_A -> CameraCommand.MoveByAnimated(KEYBOARD_PAN, 0.0, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_RIGHT,
        GLFW_KEY_D -> CameraCommand.MoveByAnimated(-KEYBOARD_PAN, 0.0, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_UP,
        GLFW_KEY_W -> CameraCommand.MoveByAnimated(0.0, KEYBOARD_PAN, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_DOWN,
        GLFW_KEY_S -> CameraCommand.MoveByAnimated(0.0, -KEYBOARD_PAN, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_EQUAL ->
          CameraCommand.ScaleByAnimated(KEYBOARD_ZOOM, viewportCenter(), KEYBOARD_ANIMATION_MS)

        GLFW_KEY_MINUS ->
          CameraCommand.ScaleByAnimated(
            1.0 / KEYBOARD_ZOOM,
            viewportCenter(),
            KEYBOARD_ANIMATION_MS,
          )

        GLFW_KEY_Q -> CameraCommand.AdjustBearingAnimated(-KEYBOARD_BEARING, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_E -> CameraCommand.AdjustBearingAnimated(KEYBOARD_BEARING, KEYBOARD_ANIMATION_MS)
        GLFW_KEY_RIGHT_BRACKET ->
          CameraCommand.AdjustPitchAnimated(KEYBOARD_PITCH, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_LEFT_BRACKET ->
          CameraCommand.AdjustPitchAnimated(-KEYBOARD_PITCH, KEYBOARD_ANIMATION_MS)

        GLFW_KEY_0 -> CameraCommand.ResetOrientation(RESET_ANIMATION_MS)
        else -> null
      }
    if (command != null) {
      commands.push(command)
      renderRequest.set()
    }
  }

  private fun viewportCenter(): ScreenPoint {
    val current = viewport()
    return ScreenPoint(current.width() / 2.0, current.height() / 2.0)
  }

  override fun close() {
    glfwSetCursorPosCallback(window, null)
    glfwSetMouseButtonCallback(window, null)
    glfwSetScrollCallback(window, null)
    glfwSetKeyCallback(window, null)
  }

  internal companion object {
    private const val DRAG_ROTATE_FACTOR = 0.5
    private const val DRAG_PITCH_FACTOR = 0.5
    private const val KEYBOARD_PAN = 120.0
    private const val KEYBOARD_ZOOM = 1.25
    private const val KEYBOARD_BEARING = 10.0
    private const val KEYBOARD_PITCH = 5.0
    private const val KEYBOARD_ANIMATION_MS = 160.0
    private const val RESET_ANIMATION_MS = 220.0

    fun printControls() {
      println("Controls:")
      println("  left drag: pan")
      println("  right drag or Ctrl+left drag: rotate with X, pitch with Y")
      println("  scroll: zoom at cursor")
      println("  arrows or WASD: pan")
      println("  + / -: zoom at center")
      println("  Q / E: rotate")
      println("  ] / [: pitch")
      println("  0: reset pitch and bearing")
    }
  }
}
